Fix URL Trailing Slash Flicker

randychilau opened the pull request:
**Description**

This PR fixes the issue of a flickering trailing slash with the initial loading of layer5.io url.

Currently:

![undesired_flicker](https://user-images.githubusercontent.com/90356410/234205428-82b1cf7f-7925-4843-906f-7ce8b422f678.gif)

After fix (showing on a [test site](https://randychilau.github.io/layer5-test) hosted on GitHub Pages):

![url_flicker_fix](https://user-images.githubusercontent.com/90356410/234207821-7658cca2-1c61-4b50-9243-319e383696a2.gif)


**Notes for Reviewers**

The issue happens because of the way GitHub Pages handles no-trailing-slash requests. Unfortunately GitHub Pages does not offer a way to change it. This can be fixed by changing some gatsby-config.js settings, fork and customize a currently used  plugin (gatsby-plugin-meta-redirect), and adding code in gatsby-node.js to only run when the build is hosted on GitHub Pages.

Note: The preview and build generated by GitHub Actions is hosted on netlify whose default settings [handle trailing slash requests](https://github.com/slorber/trailing-slash-guide/blob/main/docs/Hosting-Providers.md) are not identical to GitHub Pages, so it is better to use the[ test site](https://randychilau.github.io/layer5-test/) for review. Please feel free to ask me any questions regarding this.

There are more details on the issue and fix in this [google doc.](https://docs.google.com/document/d/1mVyGLGGPNK-yKbIZDvMN7KUihsPld2ZR-p8kpIxjI_A/edit?usp=sharing)

Since this issue happens when the build is hosted on GitHub Pages, but not during local development (for `npm start` and `npm run build`commands). The fix should only run when GitHub Actions uses the `build` command. 

For now, we will ask contributors to use another build script `buildLocal` that includes an environment variable when they want to view a build locally. If someone else knows a better method, I will be happy to help implement.
     
**[Signed commits](https://github.com/layer5io/layer5/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 

<!--
Thank you for contributing to Layer5 projects! 

Contributing Conventions:

1. Include descriptive PR titles with [<component-name>] prepended.
2. Build and test your changes before submitting a PR. 
3. Sign your commits

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.
-->

l5io commented:
🚀 Preview for commit e9ae87a8be8d34bcb5009de5eccef134a9b003b2 at: https://644790adb77ea627597999c6--layer5.netlify.app

l5io commented:
🚀 Preview for commit b8e6982b7a4c9ee622409882acdc958346d23599 at: https://6447982c34d6482f15fb3419--layer5.netlify.app

randychilau commented:
re: 
> For now, we will ask contributors to use another build script buildLocal that includes an environment variable when they want to view a build locally. If someone else knows a better method, I will be happy to help implement.

Updated to use CI env variable for detecting when built locally or via GitHub Actions with regular` build` command. This removes need to use custom `buildLocal` script and change any steps in contributors workflow.

l5io commented:
🚀 Preview for commit 26824a6f9d18958490f1641751fca17ba27eb3f3 at: https://644825e309512404f9ad666a--layer5.netlify.app

l5io commented:
🚀 Preview for commit 07e8cb17529f0a3dca4ff3e86d52a8e82b92bdc9 at: https://644853265f60f13241256c7a--layer5.netlify.app

randychilau commented:
Some additional information:

The trailing slash issue is the reason why SEO/Performance audits will report Layer5 links as redirects, even though the page loads in the browser and the resulting url matches. (Note: search engines will generally follow redirects when evaluating links and use signals from sitemap and canonical tags to determine which links are important)

Ayush mentioned [the redirect issue](https://layer5io.slack.com/archives/C015QJKUMPU/p1680782490803989) in `#websites`:

> The issue semrush audit show with the pages is "Redirect" .It appears strange to me as the pages are still accessible with the same url and not being redirected anywhere .

![image](https://user-images.githubusercontent.com/90356410/234423845-917c0403-bd49-4ee9-bcad-9734666bae53.png)

If you take a current Layer5.io url, such as `https://layer5.io/blog` from the above image, and use a [source code viewer ](https://codebeautify.org/source-code-viewer)to  see what code loads at that url, you will see:
![image](https://user-images.githubusercontent.com/90356410/234424625-6d82bda3-6caa-43ef-a387-571248019677.png)

This is a page that is generated by GitHub Pages, and it then redirects the browser to `https://layer5.io/blog/` where the page loads, and then there is some custom code that removes the trailing slash, so the url appears in the end as `https://layer5.io/blog`. You can also see this in the browser console:
![image](https://user-images.githubusercontent.com/90356410/234427729-b01b535e-3c11-4c4b-aa5e-9e42ca3215b7.png)

----

This fix will let browser requests to layer5 urls with no-trailing-slash, such as `https://layer5.io/blog`, load correctly without redirects. This is important since we are directing search engines to no-trailing-slash links via the [Layer5 sitemap](https://layer5.io/sitemap/sitemap-0.xml).

Using the [test site](https://randychilau.github.io/layer5-test) as an example in a source code viewer, we see the following for the /blog page:
![image](https://user-images.githubusercontent.com/90356410/234427165-2b3b8528-69c2-46e3-90a2-73df9823a675.png)

And if you check the network tab in the browser console, it will show 200 status for loading the page:
![image](https://user-images.githubusercontent.com/90356410/234427517-cebcb670-20f4-45c9-a1c3-26fe2bf21c2c.png)

Thanks!

Nikhil-Ladha reviewed (commented):
This is a well researched, documented effort. Kudos 👏🏼 
A coupe of points to consider:
1. The URL in the sitemap for learning paths seems to be a bit wrong, though the page renders on correct link. Check `https://randychilau.github.io/layer5-testlearn/learning-paths/mastering-service-meshes-for-developers` in sitemap. Probably, some modification needed in gatsby-node
2. The fork and updation of a public repo is a good idea, but the suggestion is move that fork to `layer5io` org, with you having appropriate credits and permissions for it so that it is sustainable. @Chadha93 or @leecalcote can help you see through it.
3. A better documentation of the current code (updated code) vs the one that needs to be done if we change to a different provider is required in the google doc.
4. Side question, is `CI` a recognised env variable? What all env are considered for `CI` to be set as `true`?

In all, it is a great work. But, I would prefer *all* the above points to be taken care of before we move ahead with this

ayushthe1 reviewed (commented):
Thanks for such a detailed research @randychilau  .It really helped me understand the real problem and cause behind it. :+1: 
Could you please explain the use of `process.env.CI` and why we need it.

ayushthe1 commented:
Also @randychilau ,just out of curiosity ,if we find a way to add a trailing slash `/` to all pages in the sitemap ;will that solve the problem ?

l5io commented:
🚀 Preview for commit 1ea5c9a31cfec7369cc5fce3235da37708a23a1b at: https://644992830c9d7a68f9f08da6--layer5.netlify.app

randychilau commented:
Thanks to everyone for taking the time to review and provide feedback.

@Nikhil-Ladha 
> The URL in the sitemap for learning paths seems to be a bit wrong, though the page renders on correct link. Check https://randychilau.github.io/layer5-testlearn/learning-paths/mastering-service-meshes-for-developers in sitemap. Probably, some modification needed in gatsby-node

Nice find! This has been resolved in the latest commit ([check test site](https://randychilau.github.io/layer5-test/sitemap/sitemap-0.xml)).
![image](https://user-images.githubusercontent.com/90356410/234702013-d58718d7-012e-4c70-a2b2-bbd081813615.png)

----

> The fork and updation of a public repo is a good idea, but the suggestion is move that fork to layer5io org, with you having appropriate credits and permissions for it so that it is sustainable. @Chadha93 or @leecalcote can help you see through it.

I moved the code from the fork into layer5io org, uninstalled the forked plugin, and added credit/attribution in the following manner: 

- For files, I created a folder `third_party` for any outside files that are referenced and included the [license in the file comments](https://github.com/layer5io/layer5/pull/4133/files#diff-720df1c8d5cd233bf36469549b3384dcf792cef47a6d070b792b3dbfb1b00a30). 
- For snippets (e.g. in the gatsby-node.js file), I included the license in comments above the code used (but not sure how to specify original owners code vs any added code).
![image](https://user-images.githubusercontent.com/90356410/234702371-11cd6c5d-1aa2-4df9-b68d-38c50ef3bd93.png)


Note: Unsure what the best practice is for adding third-party open source files/snippets or how to distinguish which parts are original owners, look forward to learning the correct method.

----

> better documentation of the current code (updated code) vs the one that needs to be done if we change to a different provider is required in the google doc.

Added more documentation on the code changes in this PR in the [doc](https://docs.google.com/document/d/1mVyGLGGPNK-yKbIZDvMN7KUihsPld2ZR-p8kpIxjI_A/edit#bookmark=id.rzbqv71l4jt4).

Regarding code for:
> if we change to a different provider

If I am understanding the above statement correctly, it would depend [on the provider](https://github.com/slorber/trailing-slash-guide/blob/main/docs/Hosting-Providers.md), and most likely they would have an option to do server side redirects either through a .htaccess file or a setting to resolve the issue without needing to use Gatsby's APIs. GitHub Pages is unique in that it provides no option to customize server side redirects.

----

> Side question, is CI a recognised env variable? What all env are considered for CI to be set as true?

> Could you please explain the use of process.env.CI and why we need it.

Since the issue only occurs on production builds hosted by GitHub Pages and not local development environments, the fix needed to be conditional. 

This conditional `process.env.NODE_ENV === "production"` did not work because it would be `true` for both cases. 

From my understanding, GitHub provides a default environmental variable called [CI](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables) that can be accessed by the code during GitHub Actions workflow which would not be present during builds on local environments, so it is possible to use process.env.CI for the conditional. More information about the [CI env practice here](https://github.com/orgs/community/discussions/26628).

@ayushthe1 

> if we find a way to add a trailing slash / to all pages in the sitemap ;will that solve the problem ?

It is possible to have a trailing slash on all pages in the sitemap which is Gatsby's default mode. In this scenario, GitHub Pages would do the following:

- requests with trailing slash will load the page
- requests with no-trailing-slash will redirect to a trailing slash path which will load the page.

I was informed by Lee:
[![image](https://user-images.githubusercontent.com/90356410/234673538-4cd9f261-d9da-402d-8a85-4bd770744e7d.png)](https://layer5io.slack.com/archives/C015QJKUMPU/p1680816345072049?thread_ts=1680782490.803989&cid=C015QJKUMPU)

So I looked for a coding solution/options with that goal in mind.

----

Notes: 
- Since the preview provided is [hosted on netlify](https://644992830c9d7a68f9f08da6--layer5.netlify.app/), we do not get to see a real "preview" of the changes on GitHub Pages, this is why I initially created a [test site](https://randychilau.github.io/layer5-test/) hosted on GitHub Pages.
- If there are issues after merging, it will impact access to pages on the site.
- If there is a decision to merge this PR, I would suggest that it be done when expected site traffic is low to minimize impact in case there are any issues and it needs to be rolled back.
- To check that the changes are working correctly, I would suggest:
  - try loading/refreshing multiple layer5.io links without-trailing-slash in the browser to confirm there is no trailing slash flicker and 200 status code in the network tab of the browser console.
![image](https://user-images.githubusercontent.com/90356410/234713055-30f540d2-114b-497a-93be-f1fbf7594165.png)
  - try loading/refreshing multiple layer5.io links with trailing-slash in the browser to confirm the page loads correctly and resulting url does not have a trailing-slash. The network tab of the browser console should show a status code 301 redirect and then status code 200
![image](https://user-images.githubusercontent.com/90356410/234713139-84d6ec97-5525-4677-9ca4-22040285b117.png)
  - checking the [sitemap url](https://layer5.io/sitemap/sitemap-0.xml) to confirm no-trailing-slash links
  - use [source code viewer](https://codebeautify.org/source-code-viewer) to view pages with and without trailing slash (you should see redirect script for urls with trailing slash
![image](https://user-images.githubusercontent.com/90356410/234711736-6a4b1cb8-842e-4bb3-9d95-5f768a08fa44.png)
  - use source code viewer to view manually created redirects (e.g. https://layer5.io/books), this should show GitHub created redirect.
![image](https://user-images.githubusercontent.com/90356410/234711575-9f8793f8-ddd3-4212-b937-1627bd15e468.png)

Cheers,
Randy

Nikhil-Ladha commented:
@randychilau what I meant with `but the suggestion is move that fork to layer5io org` was that currently you have forked the repo under your account and exporting a npm package from there, but instead what we should we doing is fork it under [Layer5-labs](https://github.com/layer5labs) org and export a npm package from there and use it. We don't need those exported functions in the gatsby-node.js file, just the repo location needs to be updated.

The `learning-paths` are still a mess 😓 in the latest commit. The TOC is not getting generated now, and the `Back` button is not working due to incorrect slashes. Check this link: https://644992830c9d7a68f9f08da6--layer5.netlify.app/learn/learning-paths/mastering-service-meshes-for-developers/introduction-to-service-meshes/istio/deploy-an-application

The doc looks good, just for a better ref to the changes made and a link to this PR in the doc.

l5io commented:
🚀 Preview for commit 76986874dba39f64c2a81293b5240f45d64e3ecf at: https://644c1224d647ff067634cb11--layer5.netlify.app

randychilau commented:
Hi @Nikhil-Ladha,

Thanks for clarifying, I completely misunderstood the `move that fork to layer5io org`! 

Happy to work with  @Chadha93 or @leecalcote and await their guidance on next steps to do the following:
> fork it under [Layer5-labs](https://github.com/layer5labs) org and export a npm package from there and use it.

In the meantime, feel free to let me know if I should return the PR to using my own fork or leave the code as is right now in the site src code.

----

Regarding:
> The learning-paths are still a mess 😓 in the latest commit. The TOC is not getting generated now, and the Back button is not working due to incorrect slashes. 

I rolled back the changes on the learning-path slugs, and instead handled the issue in the `gatsby-config.js` file in the settings for `gatsby-plugin-sitemap` which checks for proper slashes.

from [latest netlify preview](https://644c1224d647ff067634cb11--layer5.netlify.app/sitemap/sitemap-0.xml):
![image](https://user-images.githubusercontent.com/90356410/235228109-4f4cc6eb-28f1-45c4-a960-f2691c93d4ca.png)

TOC generated and back button working:
![learning-paths](https://user-images.githubusercontent.com/90356410/235228908-8f1421d0-cee0-45f5-96ac-e40ca172c876.gif)

----

> just for a better ref to the changes made and a link to this PR in the doc.

I added a link to the PR in the doc as you suggested.


Always appreciate the feedback.

Cheers,
Randy

l5io commented:
🚀 Preview for commit c67b169f0d06d2b4359e4e8edd74df0ba9ce9f0c at: https://6450be3fe31c9a1ca6917189--layer5.netlify.app

l5io commented:
🚀 Preview for commit ce3cbc4ec17463a2d7da472c34d1729db004d390 at: https://64513d604ff71d14a89c4f48--layer5.netlify.app

l5io commented:
🚀 Preview for commit 54581e4375f4f71d7edaf549c84db38612543adf at: https://64514a50e3fbfc1e4ab3859c--layer5.netlify.app

l5io commented:
🚀 Preview for commit 37f6fbdaa7ce92c387a17c9b40e5371d886d8b77 at: https://645154f41614432c4c0d6910--layer5.netlify.app

l5io commented:
🚀 Preview for commit 2ba3afba7b9f9c76b95c6121b448bd349ef0235f at: https://64515f57c1f3a0009d822514--layer5.netlify.app

l5io commented:
🚀 Preview for commit b114e74895a08c9f321cbc9daaa25c8681003f7a at: https://645198be48e55c03db03a66c--layer5.netlify.app

l5io commented:
🚀 Preview for commit 464a61c1574da7692ae30d6dfed009d54ddd1e71 at: https://64520244d81aa539775adc1f--layer5.netlify.app

l5io commented:
🚀 Preview for commit 617e0959d04dc48da482b87ef3a67f5e70af2bec at: https://64520c5ec2481847f168d7bd--layer5.netlify.app

randychilau commented:
Hi @Nikhil-Ladha ,

Thanks for setting up the repo under the layerlabs org.

I installed the package from the repo url and removed the related code in `gatsby-node.js` and the `third_party` folder.

Added code to handle url related issues:
- updated SEO component so meta tags in `head` show correct url
- updated SEO component  to add canonical link meta tag to `head` (if `canonical` prop is not already provided)
- updated the .lighthouserc.js file to resolve blocking errors and provide links for page reports ([sample](https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1683096775761-10311.report.html)). 
![image](https://user-images.githubusercontent.com/90356410/235856387-59faf77d-924d-4aa3-96d2-f194781fbb71.png)
![image](https://user-images.githubusercontent.com/90356410/235856534-11b364cc-fec6-4150-9bb9-2e5426a29b56.png)


  - note: this check will continue to "fail" because the pages do not currently meet the lighthouse audit goals `assertion failed. Exiting with status code 1.`, but all the audits do run `Done running Lighthouse!`. If the team wants the lighthouse check to show "succeeded" and have it mean all audits ran successfully (and report results), we can change some settings.

Checking the [netlify hosted preview](https://64520c5ec2481847f168d7bd--layer5.netlify.app/):
- [sitemap](https://64520c5ec2481847f168d7bd--layer5.netlify.app/sitemap/sitemap-0.xml) is displaying correct information
- canonical links and other meta tags showing correct url
![image](https://user-images.githubusercontent.com/90356410/235855972-8ca57a8a-4183-4262-a8f2-0885d4d7c9d3.png)
![image](https://user-images.githubusercontent.com/90356410/235855720-e2cac067-b707-491b-b069-1fe0af12eeb2.png)
- url fix in action for no-trailing-slash  
![image](https://user-images.githubusercontent.com/90356410/235858252-a17f7f4c-af73-4d01-8158-1408ffe169a2.png)

- in order to see the trailing slash behavior, we need to use the Github Page hosted [test site](https://randychilau.github.io/layer5-test) to simulate how the live production version will behave, example: 
[`https://randychilau.github.io/layer5-test/projects/`](https://randychilau.github.io/layer5-test/projects/)
![image](https://user-images.githubusercontent.com/90356410/235858994-5eb5cd1d-3314-48a4-9076-c13b0deb9e14.png)
checking the source viewer for `https://randychilau.github.io/layer5-test/projects/` shows the refresh script (and canonical link): 
![image](https://user-images.githubusercontent.com/90356410/235859428-b791b9ae-93bc-4409-9e91-e113e7a8eae8.png)

Thanks again for your help.

Cheers,
Randy

Nikhil-Ladha commented on package-lock.json:
There are still a lot of stale references to your fork in the `package-lock.json` file. Try running `make setup-libs` and see if that helps or deleting the file and then re running `make setup-libs` to recreate the file. That should help.

Nikhil-Ladha commented on src/components/seo.js:
```diff
@@ -32,9 +32,13 @@ const SEO = ({ canonical, description,image, schemaMarkup, title,children }) =>
     description: description || defaultDescription,
     image: `${siteUrl}${image || siteMetadataImage}`,
-    url: `${siteUrl}${pathname || ""}`,
+    url: `${siteUrl}${pathname.replace(".html", "")  || ""}`,
```
Curiious as to when can. we have `.html` in the path? We shouldn't be having it on any page, is it only if we intentionally try to do so? Forced testing?

Nikhil-Ladha commented on src/components/seo.js:
```diff
@@ -77,7 +81,7 @@ const SEO = ({ canonical, description,image, schemaMarkup, title,children }) =>
           }`}
       </noscript>
-      {canonical &&   <link rel="canonical" href={canonical} />}
+      {canonicalUrl &&   <link rel="canonical" href={canonicalUrl} />}
```
Well, based on the change at line 38-41, it seems `canonicalUrl would never be `null/false` so we can remove this conditional from here and use it directly

Nikhil-Ladha commented on src/sections/Community/index.js:
```diff
@@ -17,11 +17,11 @@ import { graphql, useStaticQuery } from "gatsby";
 import { getImage } from "gatsby-plugin-image";
 import { BgImage } from "gbimage-bridge";
 
-const CommunityMember = "./Community-pictures/Lee Calcote and Oliver Gould - CTO of Buoyant.jpg";
-
```
These definitions were done outiside the component intentionally, as we don't want the imports to be inside the component and bloat up the code. Please undo this

Nikhil-Ladha commented on src/sections/Company/Layer5-statement/index.js:
```diff
@@ -6,11 +6,11 @@ import svgBackground from "../../../assets/images/homePage-images/background.svg
 import BannerSectionWrapper from "./statement.style";
 import { StaticImage } from "gatsby-plugin-image";
 
-const imgHero = "../../../assets/images/homePage-images/Lee-Calcote-Cloud-Native-Rejekts.jpg";
```
Same as above

randychilau commented:
Hi @Nikhil-Ladha,

Thanks so much for taking the time to review and the comments.

> stale references to your fork in the package-lock.json file

This was resolved as you suggested with `make setup`

![image](https://user-images.githubusercontent.com/90356410/236007651-299ec0cb-a304-4a7e-87d8-172e329d1d74.png)

----

> Curious as to when can. we have .html in the path?

I only noticed this in an earlier netlify preview when looking at the source code:

![image](https://user-images.githubusercontent.com/90356410/236001967-f1d40e10-791a-4ecf-8b9b-73dcbc1a6b7e.png)

Which is not consistent with the no-trailing-slash links for sharing and crawling. 

In the `SEO` component, when the reach router returns `pathname` in `const { pathname } = useLocation();` I think it is using the `pathname` of the file that is actually served by GitHub Pages? (which is `[path].html`) , and not `path` from the client side router that displays in the address bar.  

Pages served have `.html` in the `pathname`, except for the home page? where I believe `pathname` would be undefined which then uses the fallback`|| ""}`. I added a removal of a trailing slash for the home page even though Google says trailing slash for [domain names is irrelevant](https://johnmu.com/trailing-slash-or-not/).

----

> it seems canonicalUrl would never be null/false so we can remove this conditional from here and use it directly

Yes, corrected. Also, I did not need to use `canonicalUrl`. I thought there was an accessibility issue if the `canonical` prop was not passed down, but I was mistaken, so I used it throughout.

![image](https://user-images.githubusercontent.com/90356410/236009790-8c53badd-792b-45ef-95c9-74da81965f49.png)
![image](https://user-images.githubusercontent.com/90356410/236009890-c8bb8747-99f9-4d29-ae94-687943f57107.png)

----

> These definitions were done outiside the component intentionally, as we don't want the imports to be inside the component and bloat up the code. Please undo this

Thanks, reverted the changes. For some reason when I was looking into lighthouse, I thought there was a warning about these images and I was troubleshooting, but I cannot find/reproduce it.

Also, regarding Lighthouse. I mistakenly thought it was breaking in general for all PRs

![image](https://user-images.githubusercontent.com/90356410/236011798-4c51d0ef-8403-4a89-ad5b-d2f7e3d04f4b.png)

but it was only for this PR because the urls in `.lighthouserc.js` were no longer valid, and needed to be updated to `[path].html` urls

side note, when I was running Lighthouse in the browser dev console, it showed this (but did not indicate that it impacted the reporting):

![image](https://user-images.githubusercontent.com/90356410/236017374-e51e2bd2-5e81-4f9d-a6c0-0837f6e73bf0.png)


----

Please let me know if there are any other issues or questions, great to get feedback.

Cheers,
Randy

l5io commented:
🚀 Preview for commit dd92a914dc35c714af969f6f11a59256c4b49c85 at: https://6452b630a9e99011236f349f--layer5.netlify.app

Nikhil-Ladha commented:
> side note, when I was running Lighthouse in the browser dev console, it showed this (but did not indicate that it impacted the reporting):
> 
> ![image](https://user-images.githubusercontent.com/90356410/236017374-e51e2bd2-5e81-4f9d-a6c0-0837f6e73bf0.png)

Yeah, lighthouse reports doesn't highlight any trailing slash issue, rather it is the [semrush](https://www.semrush.com/) report that we run weekly on our site and the generated report is send to a different private mailing lists that lists the actual issue.

Nikhil-Ladha commented:
Doing a final round of checks, before approving and merging...

Nikhil-Ladha reviewed (approved):
LGTM 🚀
